### ISSUE TITLE
Job balance among cores

### ISSUE THREAD

**Issue by sunez** [avadhpatel/marss#51]
Thank you for your efforts. I found that jobs are not equally distributed among cores. For example, I run 4 threads among 4 cores. I found that 1 or 2 cores are often not used to run my programs. If I am lucky, all 4 cores are used, but it happens not that often. What I found is that if I run one marssX86 simulation at a time, normally it uses all 4 cores. However, this is not possible configuration. I have to run multiple simulations concurrently. Could you tell me what I missed and why it happens?  And how can I fix it?

**Comment by sunez** [comment]
By the way, if this is MarssX86 internal bug, it is a really critical one as a simulator. All the works with MarssX86 should not be trusted, as long as they used multi core configuration (but what is not multi-core nowadays). What I experienced is that it is very un-predictable. Sometimes, 2 cores run. Sometimes, 3 cores run. Worst case, it seems that none of cores run user programs, so that they just stay (not deadlock), but never finish.

**Comment by fitzfitsahero** [comment]
I'm going to need way more information if you're looking for help. 

What benchmark did you run? What configuration? For how long did you run the benchmark?

It is very possible that you never got out of the initialization phase.

**Comment by sunez** [comment]
Thank you for your response. The benchmark is what I made. It is a tree data structure with inserts and deletes. I built it with 4 OOO_core (c=4) with Dramsim2. L1 and L2 is private, and L3 is shared cache. At the beginning of the program, it creates checkpoint. From the checkpoint, the program runs about 30 minutes. 

How long does it take to get out of the initialization phase?

**Comment by fitzfitsahero** [comment]
Does your benchmark use pthreads or openmp or some other threading library?

**Comment by sunez** [comment]
Yes, it uses pthreads. Can it be a problem? If it is, could you recommend any other alternatives?

**Comment by fitzfitsahero** [comment]
We have images with parsec and splash, I would suggest you try them first. 

I have run multicore simulations that effectively use all of the cores and get expected IPCs

**Comment by sunez** [comment]
Can I send you one of my benchmarks and you can look at it? It is fairly simple codes with pthreads. I don't know how parsec is working, but I am experiencing this issue for all my benchmarks.

**Comment by sunez** [comment]
Then, my question is how many thread have you run when you do your tests? As I know PARSEC creates quite numbers of threads.  Do you know the details?

**Comment by fitzfitsahero** [comment]
I run as many threads as I have cores.

**Comment by sunez** [comment]
Could you explain how the jobs are allocated to cores? Are they done by MarssX86? Which part of codes are doing it?

**Comment by sunez** [comment]
I just checked the posted parsec image and it seems that it uses thread-affinity. In MarssX86, if the thread uses thread-affinity, only one simulations is successfully running and all others are not proceeding actually (although it looks it does). I also tried thread-affinity in my benchmarks previously and I checked this issue already.
